magenta/music/musicxml_parser.py

```diff
@@ -0,0 +1,516 @@
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#    http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+"""MusicXML parser.
+
+Simple MusicXML parser used to convert MusicXML
+into tensorflow.magenta.NoteSequence.
+"""
+
+# Imports
+# Python 2 uses integer division for integers. Using this gives the Python 3
+# behavior of producing a float when dividing integers
+from __future__ import division
+
+from zipfile import ZipFile
+import xml.etree.ElementTree as ET
+
+# Constants
+
+# Number of MIDI ticks per beat. Required by NoteSequence structure
+TICKS_PER_BEAT = 960
+
+# Global variables
+# Default to one division per measure
+current_divisions = 1
+
+# Default to a tempo of 120 beats per minute
+# (This is the default tempo according to the
+# Standard MIDI Files 1.0 Specification)
+current_tempo = 120
+
+# Duration of a single beat in seconds
+current_seconds_per_beat = 0.5
```

Comment:
I don't have a good solution for this at the moment. I know the code isn't threadsafe because I am mutating global variables, but parsing MusicXML files can be tricky. MusicXML files (like most/all? xml files) do not guarantee order of elements, only that the elements exist. For example, let's say there is a tempo marking over beat 1 in the first measure of quarter = 96. Ideally, I'd like to know the tempo before parsing the note on beat one, but MusicXML does not guarantee this. That example is trivial, but it gets more complicated when keeping track of time and you encounter the <forward> and <backup> elements. 

I'd like to keep track of this state somehow, but my only two potential solutions were to pass  around a state object to every class, or to use globals. Since I wanted to get things working quickly, I resorted to globals. I can now refactor to something more elegant and threadsafe. Do you have any alternative solutions to this problem? If not, I'll need to pass around some object that maintains these state variables.